packages/flutter/lib/src/widgets/snapshot_widget.dart

```diff
@@ -0,0 +1,434 @@
+    _painter = value;
+    if (painter.shouldRepaint(oldPainter)) {
+      markNeedsPaint();
+    }
+    if (attached) {
+      painter.addListener(markNeedsPaint);
+    }
+  }
+
+  /// A controller that determines whether to paint the child normally or to
+  /// paint a snapshotted version of that child.
+  SnapshotController get controller => _controller;
+  SnapshotController _controller;
+  set controller(SnapshotController value) {
+    if (value == controller) {
+      return;
+    }
+    controller.removeListener(_onRasterValueChanged);
+    final bool oldValue = controller.allowSnapshotting;
+    _controller = value;
+    if (attached) {
+      controller.addListener(_onRasterValueChanged);
+      if (oldValue != controller.allowSnapshotting) {
+        _onRasterValueChanged();
+      }
+    }
+  }
+
+  /// How the snapshot widget will handle platform views in child layers.
+  SnapshotMode get mode => _mode;
+  SnapshotMode _mode;
+  set mode(SnapshotMode value) {
+    if (value == _mode) {
+      return;
+    }
+    _mode = value;
+    markNeedsPaint();
+  }
+
+  ui.Image? _childRaster;
+  // Set to true if the snapshot mode was not forced and a platform view
+  // was encountered while attempting to snapshot the child.
+  bool _disableSnapshotAttempt = false;
+
+  @override
+  void attach(covariant PipelineOwner owner) {
+    controller.addListener(_onRasterValueChanged);
+    painter.addListener(markNeedsPaint);
+    super.attach(owner);
+  }
+
+  @override
+  void detach() {
+    controller.removeListener(_onRasterValueChanged);
+    painter.removeListener(markNeedsPaint);
+    _childRaster?.dispose();
+    _childRaster = null;
+    super.detach();
+  }
+
+  @override
+  void dispose() {
+    controller.removeListener(_onRasterValueChanged);
+    painter.removeListener(markNeedsPaint);
+    _childRaster?.dispose();
+    _childRaster = null;
+    super.dispose();
+  }
+
+  void _onRasterValueChanged() {
+    _childRaster?.dispose();
+    _childRaster = null;
+    markNeedsPaint();
+  }
+
+  // Paint [child] with this painting context, then convert to a raster and detach all
+  // children from this layer.
+  ui.Image? _paintAndDetachToImage() {
+    final OffsetLayer offsetLayer = OffsetLayer();
+    final PaintingContext context = PaintingContext(offsetLayer, Offset.zero & size);
+    super.paint(context, Offset.zero);
+    // This ignore is here because this method is protected by the `PaintingContext`. Adding a new
+    // method that performs the work of `_paintAndDetachToImage` would avoid the need for this, but
+    // that would conflict with our goals of minimizing painting context.
+    // ignore: invalid_use_of_protected_member
+    context.stopRecordingIfNeeded();
+    if (mode != SnapshotMode.forced && !offsetLayer.supportsRasterization()) {
+      if (mode == SnapshotMode.normal) {
+        throw FlutterError('SnapshotWidget used with a child that contains a PlatformView.');
+      }
+      _disableSnapshotAttempt = true;
```

Comment:
There doesn't appear to be a path that ever resets this to false? Is that expected? There are no conditions where we can try snapshotting again?

Reply:
Perhaps on detach we can reset the value?

Reply:
and also on reset

Reply:
or just on notification from the controller